Mathlib/RingTheory/AdicCompletion/Basic.lean

```diff
@@ -202,24 +202,57 @@ instance : Neg (AdicCompletion I M) where
 instance : Sub (AdicCompletion I M) where
   sub x y := ⟨x.val - y.val, by simp [x.property, y.property]⟩
 
-instance : SMul ℕ (AdicCompletion I M) where
-  smul n x := ⟨n • x.val, by simp [x.property]⟩
+instance instSMul [SMul S R] [SMul S M] [IsScalarTower S R M] : SMul S (AdicCompletion I M) where
+  smul r x := ⟨r • x.val, by simp [x.property]⟩
+
+@[simp, norm_cast] lemma val_zero : (0 : AdicCompletion I M).val = 0 := rfl
+
+lemma val_zero_apply (n : ℕ) : (0 : AdicCompletion I M).val n = 0 := rfl
+
+variable {I M}
+
+@[simp, norm_cast] lemma val_add (f g : AdicCompletion I M) : (f + g).val = f.val + g.val := rfl
+@[simp, norm_cast] lemma val_sub (f g : AdicCompletion I M) : (f - g).val = f.val - g.val := rfl
+@[simp, norm_cast] lemma val_neg (f : AdicCompletion I M) : (-f).val = -f.val := rfl
```

Comment:
Do you think that instead of all this API it's better to make `AdicCompletion.val : AdicCompletion I M ->+ ((n : ℕ) → M ⧸ I ^ n • ⊤)`?

Reply:
I think its usually better to have the unbundled spelling as the preferred form, and have the morphism around as well for assembling the non-refl proofs like `val_sum`.

Reply:
Yes, I would rather keep `val` unbundled here